pydantic/fields.py

```diff
@@ -737,7 +738,25 @@ def _validate_mapping(
         if errors:
             return v, errors
         else:
-            return result, None
+            return self._get_mapping_value(v, result), None
+
+    def _get_mapping_value(self, original: T, converted: Dict[Any, Any]) -> Union[T, Dict[Any, Any]]:
+        target_type = get_origin(self.outer_type_)
+        original_type = type(original)
+
+        if is_mapping_type(target_type):
+            target_type = original_type
+
+        if target_type is dict:
+            return converted
+        elif target_type is defaultdict:
+            return defaultdict(getattr(original, 'default_factory', None), **converted)
```

Comment:
```suggestion
            return defaultdict(getattr(original, 'default_factory', None), converted)
```
and same under.
If we use `kwargs`, it won't work for `int` keys. But it does when passing a dict as `arg`.
Same under